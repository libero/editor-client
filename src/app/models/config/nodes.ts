export const nodes = {
  doc: {
    content: 'block'
  },

  'article-title': {
    group: 'block',
    content: 'inline*',
    parseDom: [{ tag: 'div', attrs: {'data-tag': 'article-title'}}],
    toDOM(node) {
      return ['div', {'data-tag': 'article-title', class: 'article-title'}, 0]
    }
  },

  'kwd-group': {
    group: 'block',
    content: 'block*',
    parseDom: [{ tag: 'div', attrs: {'data-tag': 'keyword-group'}}],
    toDOM(node) {
      return ['div', {'data-tag': 'keyword-group', class: 'keyword-group'}, 0]
    }
  },

  'kwd': {
    group: 'block',
    content: 'inline*',
    parseDom: [{ tag: 'div', attrs: {'data-tag': 'keyword'}}],
    toDOM(node) {
      return ['div', {'data-tag': 'keyword', class: 'keyword'}, 0]
    }
  },

  text: {
    group: 'inline'
  }
};

export type NodesType = keyof typeof nodes;
