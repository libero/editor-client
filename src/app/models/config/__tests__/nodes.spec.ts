import { Node as ProsemirrorNode } from 'prosemirror-model';
import { nodes } from '../nodes';

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

jest.mock('app/models/figure', () => ({
  getFigureImageUrlFromXml: () => 'IMAGE_URL',
  createFigureLicenseAttributes: () => ({
    copyrightHolder: 'Copyright holder',
    copyrightStatement: 'Copyright stmt',
    copyrightYear: 'Copyright year',
    licenseType: 'License type'
  }),
  createEmptyLicenseAttributes: () => ({
    copyrightHolder: '',
    copyrightStatement: '',
    copyrightYear: '',
    licenseType: ''
  })
}));

describe('nodes spec', () => {
  it('checks nodes spec definition', () => {
    expect(nodes).toMatchSnapshot();
  });

  it('checks annotatedReferenceInfoDoc spec', () => {
    expect(nodes['annotatedReferenceInfoDoc'].toDOM()).toEqual(['p', 0]);
  });

  it('checks article-title spec', () => {
    expect(nodes['article-title'].toDOM()).toEqual(['h1', { class: 'article-title' }, 0]);
  });

  it('checks abstract spec', () => {
    expect(nodes['abstract'].toDOM()).toEqual(['p', { class: 'abstract' }, 0]);
  });

  it('checks keyword spec', () => {
    expect(nodes['keyword'].toDOM()).toEqual(['div', 0]);
  });

  it('checks paragraph spec', () => {
    expect(nodes['paragraph'].toDOM()).toEqual(['p', { class: 'paragraph' }, 0]);
  });

  it('checks section spec', () => {
    expect(nodes['section'].toDOM()).toEqual(['section', 0]);
  });

  it('checks boxText spec', () => {
    expect(nodes['boxText'].toDOM()).toEqual(['section', { class: 'box-text' }, 0]);
  });

  it('checks figureTitle spec', () => {
    expect(nodes['figureTitle'].toDOM()).toEqual(['p', 0]);
  });

  it('checks figureLegend spec', () => {
    expect(nodes['figureLegend'].toDOM()).toEqual(['p', 0]);
  });

  it('checks figureAttribution spec', () => {
    expect(nodes['figureAttribution'].toDOM()).toEqual(['p', 0]);
  });

  it('checks orderedList spec', () => {
    expect(nodes['orderedList'].toDOM()).toEqual(['ol', 0]);
  });

  it('checks bulletList spec', () => {
    expect(nodes['bulletList'].toDOM()).toEqual(['ul', 0]);
  });

  it('checks listItem spec', () => {
    expect(nodes['listItem'].toDOM()).toEqual(['li', 0]);
  });

  it('checks heading spec attributes', () => {
    const body = document.createElement('body');
    body.innerHTML = `<body><sec>
        <title id="title1">Test</title>
        <sec>
            <title id="title2"> Second level heading </title>
        </sec>
    </sec><body>`;
    expect(nodes['heading'].parseDOM[0].getAttrs(body.querySelector('#title1'))).toEqual({
      domId: 'unique_id',
      level: 1
    });
    expect(nodes['heading'].parseDOM[0].getAttrs(body.querySelector('#title2'))).toEqual({
      domId: 'unique_id',
      level: 2
    });
  });

  it('checks heading spec rendering', () => {
    const node = new ProsemirrorNode();
    node.attrs = { level: 3, domId: 'SOME_ID' };
    expect(nodes['heading'].toDOM(node)).toEqual(['h3', { id: 'SOME_ID' }, 0]);
  });

  it('checks refCitation spec XML attributes', () => {
    const node = document.createElement('xref');
    node.innerHTML = 'SOME_TEXT';
    node.setAttribute('rid', 'SOME_ID');
    expect(nodes['refCitation'].parseDOM[0].getAttrs(node)).toEqual({ refText: 'SOME_TEXT', refId: 'SOME_ID' });
  });

  it('checks refCitation spec HTML (clipboards) attributes', () => {
    const node = document.createElement('a');
    node.innerHTML = 'SOME_TEXT';
    node.setAttribute('data-cit-type', 'reference');
    node.setAttribute('data-ref-id', 'SOME_ID');
    node.setAttribute('data-ref-text', 'SOME_TEXT');
    expect(nodes['refCitation'].parseDOM[1].getAttrs(node)).toEqual({ refText: 'SOME_TEXT', refId: 'SOME_ID' });
  });

  it('checks refCitation spec HTML (clipboards) rendering', () => {
    const node = new ProsemirrorNode();
    node.attrs = { refId: 'SOME_ID', refText: 'SOME_TEXT' };
    expect((nodes['refCitation'].toClipboardDOM(node) as HTMLElement).outerHTML).toEqual(
      '<a href="#" data-cit-type="reference" data-ref-id="SOME_ID" data-ref-text="SOME_TEXT" class="citation">SOME_TEXT</a>'
    );
  });

  it('checks refCitation spec text (clipboards) rendering', () => {
    const node = new ProsemirrorNode();

    node.attrs = { refId: 'SOME_ID', refText: 'SOME_TEXT' };
    expect(nodes['refCitation'].toClipboardText(node)).toEqual('SOME_TEXT');
  });

  it('checks refCitation spec rendering', () => {
    const node = new ProsemirrorNode();

    node.attrs = { refId: 'SOME_ID', refText: 'SOME_TEXT' };
    expect(nodes['refCitation'].toDOM(node)).toEqual([
      'a',
      {
        href: '#',
        class: 'citation',
        'data-cit-type': 'reference',
        'data-ref-id': 'SOME_ID',
        'data-ref-text': 'SOME_TEXT'
      }
    ]);
  });

  it('checks figure spec rendering', () => {
    const node = document.createElement('fig');
    node.setAttribute('id', 'SOME_ID');
    node.setAttribute('position', 'SOME_POS');
    node.innerHTML = '<label>SOME_LABEL</label>';
    expect(nodes['figure'].parseDOM[0].getAttrs(node)).toEqual({
      id: 'SOME_ID',
      label: 'SOME_LABEL',
      img: 'IMAGE_URL',
      position: 'SOME_POS'
    });
  });

  it('checks figureLicese attributes', () => {
    const node = document.createElement('fig');
    expect(nodes['figureLicense'].parseDOM[0].getAttrs(node)).toEqual({
      licenseInfo: {
        copyrightHolder: 'Copyright holder',
        copyrightStatement: 'Copyright stmt',
        copyrightYear: 'Copyright year',
        licenseType: 'License type'
      }
    });
  });

  it('checks figureLicese rendering', () => {
    expect(nodes['figureLicense'].toDOM()).toEqual(['p', 0]);
  });

  it('checks figureCitation spec XML attributes', () => {
    const node = document.createElement('xref');
    node.setAttribute('rid', 'SOME_ID');
    expect(nodes['figureCitation'].parseDOM[0].getAttrs(node)).toEqual({ figIds: ['SOME_ID'] });
  });

  it('checks figureCitation spec HTML (clipboards) attributes', () => {
    const node = document.createElement('a');
    node.setAttribute('data-fig-ids', 'SOME_ID1 SOME_ID2');
    expect(nodes['figureCitation'].parseDOM[1].getAttrs(node)).toEqual({ figIds: ['SOME_ID1', 'SOME_ID2'] });
  });

  it('checks figureCitation spec rendering', () => {
    const node = new ProsemirrorNode();

    node.attrs = { figIds: ['SOME_ID1', 'SOME_ID2'] };
    expect(nodes['figureCitation'].toDOM(node)).toEqual([
      'a',
      {
        href: '#',
        class: 'citation',
        'data-cit-type': 'figure',
        'data-fig-ids': 'SOME_ID1 SOME_ID2'
      },
      0
    ]);
  });
});
