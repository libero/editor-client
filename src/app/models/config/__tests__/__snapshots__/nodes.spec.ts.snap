// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodes spec checks nodes spec definition 1`] = `
Object {
  "abstract": Object {
    "content": "paragraph",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "abstract",
      },
    ],
    "toDOM": [Function],
  },
  "annotatedReferenceInfoDoc": Object {
    "content": "inline*",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "article-title",
      },
      Object {
        "tag": "source",
      },
    ],
    "toDOM": [Function],
  },
  "article-title": Object {
    "content": "inline*",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "article-title",
      },
    ],
    "toDOM": [Function],
  },
  "boxText": Object {
    "content": "paragraph*",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "boxed-text",
      },
    ],
    "toDOM": [Function],
  },
  "bulletList": Object {
    "content": "listItem+",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "list[list-type=bullet]",
      },
    ],
    "toDOM": [Function],
  },
  "doc": Object {
    "content": "block+",
  },
  "figure": Object {
    "atom": true,
    "attrs": Object {
      "id": Object {
        "default": "",
      },
      "img": Object {
        "default": "",
      },
      "label": Object {
        "default": "",
      },
    },
    "content": "(figureTitle | figureLegend | figureAttribution | figureLicense)*",
    "draggable": true,
    "group": "block",
    "parseDOM": Array [
      Object {
        "getAttrs": [Function],
        "getContent": [Function],
        "tag": "fig",
      },
    ],
    "toDOM": [Function],
  },
  "figureAttribution": Object {
    "content": "inline*",
    "context": "figure",
    "parseDOM": Array [
      Object {
        "priority": 100,
        "tag": "attrib",
      },
      Object {
        "ignore": true,
        "tag": "label",
      },
    ],
    "toDOM": [Function],
  },
  "figureCitation": Object {
    "attrs": Object {
      "figIds": Object {
        "default": null,
      },
    },
    "content": "text*",
    "group": "inline",
    "inline": true,
    "parseDOM": Array [
      Object {
        "getAttrs": [Function],
        "tag": "xref[ref-type=\\"fig\\"]",
      },
      Object {
        "getAttrs": [Function],
        "priority": 100,
        "tag": "a.citation[data-cit-type=\\"figure\\"]",
      },
    ],
    "toDOM": [Function],
  },
  "figureLegend": Object {
    "content": "inline*",
    "context": "figure",
    "parseDOM": Array [
      Object {
        "priority": 100,
        "tag": "caption > p",
      },
      Object {
        "ignore": true,
        "tag": "label",
      },
    ],
    "toDOM": [Function],
  },
  "figureLicense": Object {
    "attrs": Object {
      "licenseInfo": Object {
        "default": Object {
          "copyrightHolder": "",
          "copyrightStatement": "",
          "copyrightYear": "",
          "licenseType": "",
        },
      },
    },
    "content": "inline*",
    "contentElement": "license > license-p",
    "group": "block",
    "parseDOM": Array [
      Object {
        "getAttrs": [Function],
        "tag": "permissions",
      },
      Object {
        "ignore": true,
        "tag": "label",
      },
    ],
    "toDOM": [Function],
  },
  "figureTitle": Object {
    "content": "inline*",
    "parseDOM": Array [
      Object {
        "tag": "caption > title",
      },
    ],
    "toDOM": [Function],
  },
  "heading": Object {
    "attrs": Object {
      "domId": Object {
        "default": "unique_id",
      },
      "level": Object {
        "default": 1,
      },
    },
    "content": "inline*",
    "defining": true,
    "group": "block",
    "parseDOM": Array [
      Object {
        "getAttrs": [Function],
        "tag": "sec > title",
      },
      Object {
        "attrs": Object {
          "domId": "unique_id",
          "level": 1,
        },
        "tag": "h1",
      },
      Object {
        "attrs": Object {
          "domId": "unique_id",
          "level": 2,
        },
        "tag": "h2",
      },
      Object {
        "attrs": Object {
          "domId": "unique_id",
          "level": 3,
        },
        "tag": "h3",
      },
      Object {
        "attrs": Object {
          "domId": "unique_id",
          "level": 4,
        },
        "tag": "h4",
      },
    ],
    "toDOM": [Function],
  },
  "keyword": Object {
    "atom": true,
    "content": "text*",
    "parseDOM": Array [
      Object {
        "tag": "kwd",
      },
    ],
    "toDOM": [Function],
  },
  "listItem": Object {
    "content": "paragraph block*",
    "defining": true,
    "parseDOM": Array [
      Object {
        "tag": "list-item",
      },
    ],
    "toDOM": [Function],
  },
  "orderedList": Object {
    "content": "listItem+",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "list[list-type=order]",
      },
    ],
    "toDOM": [Function],
  },
  "paragraph": Object {
    "content": "inline*",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "p",
      },
    ],
    "toDOM": [Function],
  },
  "refCitation": Object {
    "atom": true,
    "attrs": Object {
      "refId": Object {
        "default": null,
      },
      "refText": Object {
        "default": null,
      },
    },
    "content": undefined,
    "group": "inline",
    "inline": true,
    "parseDOM": Array [
      Object {
        "getAttrs": [Function],
        "tag": "xref[ref-type=\\"bibr\\"]",
      },
      Object {
        "getAttrs": [Function],
        "priority": 100,
        "tag": "a.citation[data-cit-type=\\"reference\\"]",
      },
    ],
    "toClipboardDOM": [Function],
    "toClipboardText": [Function],
    "toDOM": [Function],
  },
  "section": Object {
    "content": "block*",
    "group": "block",
    "parseDOM": Array [
      Object {
        "tag": "sec",
      },
    ],
    "toDOM": [Function],
  },
  "text": Object {
    "group": "inline",
  },
}
`;
